photos/db.go

```diff
@@ -0,0 +1,478 @@
+INSERT INTO photos.users VALUES ($1, 0, 0, $2, $3);
+`
+	const minNameLen = 1
+	const maxNameLen = 30
+	const minAddrLen = 20
+	const maxAddrLen = 100
+	name := randStringBytes(minNameLen + rand.Intn(maxNameLen-minNameLen))
+	addr := randStringBytes(minAddrLen + rand.Intn(maxAddrLen-minAddrLen))
+	_, err = tx.Exec(insertSQL, userID, name, addr)
+	return err
+}
+
+// createPhoto looks up or creates a new user to match userID (it's
+// the only method in this interface which doesn't match an existing
+// user except for createUser). It then creates a new photo for the
+// new or pre-existing user.
+func createPhoto(tx *sql.Tx, userID int) error {
+	if err := createUser(tx, userID); err != nil {
+		return err
+	}
+
+	const insertSQL = `
+INSERT INTO photos.photos VALUES (DEFAULT, $1, 0, $2, $3, $4, NOW());
+`
+	const minCaptionLen = 10
+	const maxCaptionLen = 200
+	caption := randStringBytes(minCaptionLen + rand.Intn(maxCaptionLen-minCaptionLen))
+	latitude := rand.Float32() * 90
+	longitude := rand.Float32() * 180
+	if _, err := tx.Exec(insertSQL, userID, caption, latitude, longitude); err != nil {
+		return err
+	}
+
+	const updateSQL = `
+UPDATE photos.users SET photoCount = photoCount + 1 WHERE id = $1;
+`
+	if _, err := tx.Exec(updateSQL, userID); err != nil {
+		return err
+	}
+	return nil
+}
+
+// createComment chooses a random photo from a user with the closest
+// matching user ID and generates a random author ID to author the
+// comment. Counts are updated on the photo and author user.
+func createComment(tx *sql.Tx, userID int) error {
+	photoID, err := chooseRandomPhoto(tx, userID)
+	if err != nil {
+		return err
+	}
+	authorID := rand.Intn(userID) + 1
+
+	const insertSQL = `
+INSERT INTO photos.comments VALUES ($1, DEFAULT, $2, $3, NOW());
+`
+	const minMessageLen = 32
+	const maxMessageLen = 1024
+	message := randStringBytes(minMessageLen + rand.Intn(maxMessageLen-minMessageLen))
+	if _, err := tx.Exec(insertSQL, photoID, authorID, message); err != nil {
+		log.Infof("insert into photos failed: %s", err)
+		return err
+	}
+
+	const updatePhotoSQL = `
+UPDATE photos.photos SET commentCount = commentCount + 1 WHERE id = $1;
+`
+	if _, err := tx.Exec(updatePhotoSQL, photoID); err != nil {
+		return err
+	}
+
+	const updateUserSQL = `
+UPDATE photos.users SET commentCount = commentCount + 1 WHERE id = $1;
+`
+	if _, err := tx.Exec(updateUserSQL, authorID); err != nil {
+		return err
+	}
+	return nil
+}
+
+// listPhotos queries up to 100 photos, sorted by timestamp in
+// descending order, for the first user with ID >= userID. If photoIDs
+// is not nil, stores the queried photo IDs in photoIDs.
+func listPhotos(tx *sql.Tx, userID int, photoIDs *[][]byte) error {
+	var err error
+	userID, err = findClosestUserByID(tx, userID)
+	if err != nil {
+		return err
+	}
+	const selectSQL = `
+SELECT id, caption, commentCount, latitude, longitude, timestamp FROM photos.photos WHERE userID = $1 ORDER BY timestamp DESC LIMIT 100`
```

Comment:
Added. @petermattis if I create an index as follows: `CREATE UNIQUE INDEX IF NOT EXISTS photosByUserID ON photos.photos (userID, timestamp);`, it will do the right thing with this select statement, correct?